File: packages/ts/tests/e2e/declared-table-name.test.ts
// CLI e2e: the declared `[[table]] name` key through the real launcher.
//
// A table's name is declared, never derived from `ddl`. The launcher must
// query a config table under its declared `name`, exit non-zero when a
// `[[table]]` entry omits `name`, and exit non-zero when the entry's `ddl`
// never creates that name. Nothing is mocked: real launcher, real process,
// real filesystem, real SQLite, real `on-file` command spawn.

import { spawn } from "node:child_process";
import { mkdir, mkdtemp, readFile, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PKG_ROOT = join(__dirname, "..", "..");

const PKG: { bin: { dirsql: string } } = JSON.parse(
  await readFile(join(PKG_ROOT, "package.json"), "utf8"),
);
const CLI_ENTRY = join(PKG_ROOT, PKG.bin.dirsql);

interface RunResult {
  code: number | null;
  stderr: string;
  stdout: string;
}

function runLauncher(args: string[], cwd: string): Promise<RunResult> {
  return new Promise((resolve) => {
    const proc = spawn(process.execPath, [CLI_ENTRY, ...args], {
      cwd,
      stdio: ["pipe", "pipe", "pipe"],
    });
    let stderr = "";
    let stdout = "";
    proc.stderr.setEncoding("utf8");
    proc.stdout.setEncoding("utf8");
    proc.stderr.on("data", (chunk: string) => {
      stderr += chunk;
    });
    proc.stdout.on("data", (chunk: string) => {
      stdout += chunk;
    });
    proc.stdin.end();
    proc.on("close", (code) => resolve({ code, stderr, stdout }));
    setTimeout(() => {
      if (proc.exitCode === null) {
        proc.kill("SIGKILL");
      }
    }, 20_000);
  });
}

describe("declared [[table]] name (CLI)", () => {
  let dir: string;
  let configPath: string;

  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "dirsql-declared-name-e2e-"));
    configPath = join(dir, ".dirsql.toml");
    await mkdir(join(dir, "data"), { recursive: true });
    await writeFile(
      join(dir, "data", "a.json"),
      '[{"id": "one"}, {"id": "two"}]',
    );
  });

  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  it("queries a table under its declared name", async () => {
    await writeFile(
      configPath,
      `
[[table]]
name = "records"
ddl = "CREATE TABLE records (id TEXT)"
glob = "data/*.json"
on-file = "cat {path}"
`,
    );

    const { code, stdout, stderr } = await runLauncher(
      ["query", "SELECT id FROM records ORDER BY id", "--config", configPath],
      dir,
    );

    expect(code, `expected a clean run, stderr: ${stderr}`).toBe(0);
    expect((JSON.parse(stdout) as { id: string }[]).map((r) => r.id)).toEqual([
      "one",
      "two",
    ]);
  }, 30_000);

  it("exits non-zero when a [[table]] entry has no name", async () => {
    await writeFile(
      configPath,
      `
[[table]]
ddl = "CREATE TABLE records (id TEXT)"
glob = "data/*.json"
on-file = "cat {path}"
`,
    );

    const { code, stderr } = await runLauncher(
      ["query", "SELECT id FROM records", "--config", configPath],
      dir,
    );

    expect(code).not.toBe(0);
    expect(stderr).toContain("name");
    expect(stderr).toContain("[[table]]");
  }, 30_000);

  it("exits non-zero when the ddl never creates the declared name", async () => {
    await writeFile(
      configPath,
      `
[[table]]
name = "messages"
ddl = "CREATE TABLE records (id TEXT)"
glob = "data/*.json"
on-file = "cat {path}"
`,
    );

    const { code, stderr } = await runLauncher(
      ["query", "SELECT id FROM messages", "--config", configPath],
      dir,
    );

    expect(code).not.toBe(0);
    expect(stderr).toContain("table 'messages'");
  }, 30_000);
});

File: packages/python/tests/e2e/declared_table_name_test.py
"""CLI e2e: the declared `[[table]] name` key through the real launcher.

A table's name is declared, never derived from `ddl`. The launcher must query
a config table under its declared `name`, exit non-zero when a `[[table]]`
entry omits `name`, and exit non-zero when the entry's `ddl` never creates
that name. No mocks: real console script, real process, real filesystem, real
SQLite, real `on-file` command spawn.
"""

from __future__ import annotations

import json
import shutil
import subprocess


def _cli() -> str:
    """Resolve the `dirsql` console script for this test env."""
    dirsql = shutil.which("dirsql")
    assert dirsql is not None, (
        "`dirsql` console script not on PATH -- run `uv run maturin develop`"
    )
    return dirsql


def _fixture(tmp_path, config: str):
    root = tmp_path / "data-root"
    (root / "data").mkdir(parents=True)
    (root / "data" / "a.json").write_text('[{"id": "one"}, {"id": "two"}]')
    cfg = root / ".dirsql.toml"
    cfg.write_text(config)
    return root, cfg


def _query(root, cfg, sql: str):
    return subprocess.run(
        [_cli(), "query", sql, "--config", str(cfg)],
        cwd=str(root),
        stdin=subprocess.DEVNULL,
        capture_output=True,
        text=True,
        timeout=30,
    )


def describe_declared_table_name():
    def it_queries_a_table_under_its_declared_name(tmp_path):
        root, cfg = _fixture(
            tmp_path,
            "[[table]]\n"
            'name = "records"\n'
            'ddl = "CREATE TABLE records (id TEXT)"\n'
            'glob = "data/*.json"\n'
            'on-file = "cat {path}"\n',
        )

        proc = _query(root, cfg, "SELECT id FROM records ORDER BY id")

        assert proc.returncode == 0, f"expected success, got {proc.stderr!r}"
        assert [row["id"] for row in json.loads(proc.stdout)] == ["one", "two"]

    def it_exits_nonzero_when_a_table_entry_has_no_name(tmp_path):
        root, cfg = _fixture(
            tmp_path,
            "[[table]]\n"
            'ddl = "CREATE TABLE records (id TEXT)"\n'
            'glob = "data/*.json"\n'
            'on-file = "cat {path}"\n',
        )

        proc = _query(root, cfg, "SELECT id FROM records")

        assert proc.returncode != 0, f"expected failure, got {proc!r}"
        assert "name" in proc.stderr and "[[table]]" in proc.stderr, (
            f"stderr must name the missing `name` key, got {proc.stderr!r}"
        )

    def it_exits_nonzero_when_the_ddl_never_creates_the_declared_name(tmp_path):
        root, cfg = _fixture(
            tmp_path,
            "[[table]]\n"
            'name = "messages"\n'
            'ddl = "CREATE TABLE records (id TEXT)"\n'
            'glob = "data/*.json"\n'
            'on-file = "cat {path}"\n',
        )

        proc = _query(root, cfg, "SELECT id FROM messages")

        assert proc.returncode != 0, f"expected failure, got {proc!r}"
        assert "table 'messages'" in proc.stderr, (
            f"stderr must carry the config-entry prefix, got {proc.stderr!r}"
        )
